### PR TITLE
fix(hc): Prevent exception from siloed_atomic

### PR DESCRIPTION
**Pull request by RyanSkonnord**
An exception rises from `router.db_for_write` if the database doesn't exist in the environment, as observed with `SENTRY_USE_SILOS` set to true. Introduce a custom exception type to mark this condition, and catch it when setting the atomic impl. This should be equivalent behavior, as we only care about whether "missing" database is equal to `using` (which a null value won't be).

### CODE REVIEW

**Comment by RyanSkonnord** [comment]
Related to https://github.com/getsentry/sentry/pull/51967.

@GabeVillalobos Let me know if you agree with my take on the error condition. You should be able to reproduce it by running `SENTRY_USE_SILOS=1 sentry devserver` from master. If you would expect `router.db_for_write` to return something in the (eventual) siloed production env, then maybe something about our devserver mode isn't behaving correctly.

**Review by markstory — APPROVED** [review]

**Review by dashed — APPROVED** [review]

File: src/sentry/silo/patches/silo_aware_transaction_patch.py [hunk 1]
```diff
@@ -20,24 +32,17 @@ def siloed_atomic(using: Optional[str] = None, savepoint: bool = True) -> Atomic
...
 
     if both_silos_route_to_same_db or current_silo_mode == SiloMode.MONOLITH:
         pass
-    elif (
-        using == router.db_for_write(ControlOutbox)
-        and SiloMode.get_current_mode() != SiloMode.CONTROL
-    ):
+    elif using == control_db and SiloMode.get_current_mode() != SiloMode.CONTROL:
         raise MismatchedSiloTransactionError(
             f"Cannot use transaction.atomic({using}) in Control Mode"
```
Comment (dashed) [inline src/sentry/silo/patches/silo_aware_transaction_patch.py:43]: @RyanSkonnord unsure if these error copies need to be corrected.
```suggestion
            f"Cannot use transaction.atomic({using}) in Region Mode"
```

Comment (RyanSkonnord) [reply src/sentry/silo/patches/silo_aware_transaction_patch.py:43]: Yes, seems like those should be switched. Or perhaps should say "*except* in control mode". @GabeVillalobos?

Comment (dashed) [reply src/sentry/silo/patches/silo_aware_transaction_patch.py:43]: That copy is much better 👍

Comment (RyanSkonnord) [reply src/sentry/silo/patches/silo_aware_transaction_patch.py:43]: Great. But it's preexisting, so let's handle it in a separate PR.

Comment (GabeVillalobos) [reply src/sentry/silo/patches/silo_aware_transaction_patch.py:43]: Ahh yep 🙈 That's correct. I'll tackle it in this PR #52377 that is split from some of @corps work in Pr #52334

**Pull request merged**